ami/trainers/image_vae_trainer.py

```diff
@@ -0,0 +1,56 @@
+from functools import partial
+
+import torch
+from torch.distributions import kl_divergence
+from torch.distributions.normal import Normal
+from torch.nn.functional import mse_loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from ami.data.buffers.buffer_names import BufferNames
+from ami.models.model_names import ModelNames
+from ami.models.model_wrapper import ModelWrapper
+from ami.models.vae import VAE, Decoder, Encoder
+
+from .base_trainer import BaseTrainer
+
+
+class ImageVAETrainer(BaseTrainer):
+    def __init__(
+        self,
+        partial_dataloader: partial[DataLoader[torch.Tensor]],
+        partial_optimizer: partial[Optimizer],
+        device: torch.device,
+    ):
+        super().__init__()
+        self.partial_optimizer = partial_optimizer
+        self.partial_dataloader = partial_dataloader
+        self.device = device
+
+    def on_data_users_dict_attached(self) -> None:
+        self.data_user = self.get_data_user(BufferNames.IMAGE)
+
+    def on_model_wrappers_dict_attached(self) -> None:
+        self.encoder: ModelWrapper[Encoder] = self.get_training_model(ModelNames.IMAGE_ENCODER)
+        self.decoder: ModelWrapper[Decoder] = self.get_training_model(ModelNames.IMAGE_DECODER)
+        self.vae = VAE(self.encoder.model, self.decoder.model)
+        self.optimizer_state = self.partial_optimizer(self.vae.parameters()).state_dict()
+
+    def train(self) -> None:
+        optimizer = self.partial_optimizer(self.vae.parameters())
+        optimizer.load_state_dict(self.optimizer_state)
+        dataset = self.data_user.get_new_dataset()
+        dataloader = self.partial_dataloader(dataset=dataset)
+        for batch in dataloader:
+            (image_batch,) = batch
+            image_batch = image_batch.to(self.device)
+            optimizer.zero_grad()
+            image_batch_reconstructed, dist_batch = self.vae(image_batch)
+            rec_loss = mse_loss(image_batch, image_batch_reconstructed)
+            kl_loss = kl_divergence(
+                dist_batch, Normal(torch.zeros_like(dist_batch.mean), torch.ones_like(dist_batch.stddev))
+            )
+            loss = rec_loss + kl_loss.sum()
+            loss.backward()
+            optimizer.step()
+        self.optimizer_state = self.partial_optimizer(self.vae.parameters()).state_dict()
```

Comment:
以前の学習の状態を一時保存するのでこうですね。
```suggestion
        self.optimizer_state = optimizer.state_dict()
```

Reply:
https://github.com/MLShukai/ami/pull/75/commits/42736a45d0c297feafed76f0af4184892ae83040
修正しました